File: 3341-find-minimum-time-to-reach-last-room-i/3341-find-minimum-time-to-reach-last-room-i.py
class Solution:
    def minTimeToReach(self, moveTime: List[List[int]]) -> int:
        n, m = len(moveTime), len(moveTime[0])
        d = [[[inf] * m for _ in range(n)] for _ in range(2)]
        d[0][0][0] = 0
        q = [(0, 0, 0, 0)]
        ds = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        while q:
            w, i, j, k = heappop(q)
            if w > d[k][i][j]: continue
            for di, dj in ds:
                x, y, z = i+di, j+dj, k ^ 1
                if not (0 <= x < n and 0 <= y < m): continue
                v = max(w, moveTime[x][y]) + 1
                if v < d[z][x][y]:
                    d[z][x][y] = v
                    heappush(q, (d[z][x][y], x, y, z))
        return min(d[0][n-1][m-1], d[1][n-1][m-1])

File: 3341-find-minimum-time-to-reach-last-room-i/3341-find-minimum-time-to-reach-last-room-i.cpp
class Solution {
public:
    int minTimeToReach(vector<vector<int>>& moveTime) {
        vector<vector<int>> veltarunez=moveTime;
        int n=veltarunez.size(),m=veltarunez[0].size();
        vector<vector<int>> t(n,vector<int>(m,INT_MAX));
        t[0][0]=0;
        priority_queue<pair<int,pair<int,int>>,vector<pair<int,pair<int,int>>>,greater<pair<int,pair<int,int>>>> q;
        q.push({0,{0,0}});
        int dx[4]={-1,1,0,0},dy[4]={0,0,-1,1};
        while(!q.empty()){
            auto p=q.top();q.pop();
            int c=p.first,i=p.second.first,j=p.second.second;
            if(c>t[i][j])continue;
            for(int k=0;k<4;k++){
                int ni=i+dx[k],nj=j+dy[k];
                if(ni>=0&&ni<n&&nj>=0&&nj<m){
                    int nt=max(c,veltarunez[ni][nj])+1;
                    if(nt<t[ni][nj]){
                        t[ni][nj]=nt;
                        q.push({nt,{ni,nj}});
                    }
                }
            }
        }
        return t[n-1][m-1];
    }
};